addons/web/controllers/binary.py

```diff
@@ -84,30 +84,58 @@ def content_common(self, xmlid=None, model='ir.attachment', id=None, field='raw'
         res.headers['Content-Security-Policy'] = "default-src 'none'"
         return res
 
-    @http.route(['/web/assets/debug/<string:filename>',
-        '/web/assets/debug/<path:extra>/<string:filename>',
-        '/web/assets/<int:id>/<string:filename>',
-        '/web/assets/<int:id>-<string:unique>/<string:filename>',
-        '/web/assets/<int:id>-<string:unique>/<path:extra>/<string:filename>'], type='http', auth="public")
+    @http.route([
+        '/web/assets/<path:extra>/<string:filename>',
+        '/web/assets/<string:unique>/<path:extra>/<string:filename>'], type='http', auth="public")
     # pylint: disable=redefined-builtin,invalid-name
-    def content_assets(self, id=None, filename=None, unique=False, extra=None, nocache=False):
-        if not id:
-            domain = [('url', '!=', False)]
-            if extra:
-                domain += [('url', '=like', f'/web/assets/%/{extra}/{filename}')]
-            else:
-                domain += [
-                    ('url', '=like', f'/web/assets/%/{filename}'),
-                    ('url', 'not like', f'/web/assets/%/%/{filename}')
-                ]
+    def content_assets(self, filename=None, unique='%', extra='-', nocache=False):
+        debug_assets = unique == 'debug'
+        attachment = None
+        if unique != 'debug':
+            url = f'/web/assets/{unique}/{extra}/{filename}'
+            domain = [
+                ('public', '=', True),
+                ('url', '!=', False),
+                ('url', '=like', url),
+                ('res_model', '=', 'ir.ui.view'),
+                ('res_id', '=', 0),
+            ]
             attachment = request.env['ir.attachment'].sudo().search(domain, limit=1)
-            if not attachment:
+        if not attachment:
+            # try to generate one
+            extra_parts = extra.split('_')
+            try:
+                params, rtl = request.env['ir.asset']._parse_assets_extra(extra_parts)
+            except ValueError:
+                _logger.exception('Invalid assets')
                 raise request.not_found()
-            id = attachment.id
+            if debug_assets:
+                bundle_name, asset_type = filename.rsplit('.', 1)
+            else:
+                bundle_name, _min, asset_type = filename.rsplit('.', 2)
+                assert _min == 'min'
+            css = asset_type == 'css'
+            js = asset_type == 'js'
+            assert js or css
```

Comment:
I don't like nor this 404 nor those unsafe rsplit nor those asserts, in case the given filename/extra are bad it is going to fail with a 500 Internal Server error but it is the request that is bad, not the server.

I would like to suggest to move the entire "try to generate one" block inside of a dedicated method to reraise any ValueError as BadRequest:

```py
if not attachment:
    with replace_exception(ValueError, by=werkzeug.exceptions.BadRequest):
        attachment = self._try_to_generate_one(...)  # todo: better name
```